sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -23,6 +23,16 @@ import org.apache.spark.sql.catalyst.trees.{CurrentOrigin, TreeNode, TreeNodeTag
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types.{DataType, StructType}
 
+/**
+ * An abstraction of the Spark SQL query plan tree, which can be logical or physical. This class
+ * defines some basic properties of a query plan node, as well as some new transform APIs to
+ * transform the expressions of the plan node.
+ *
+ * Note that, the query plan is a mutually recursive structure:
+ *   QueryPlan -> Expression (subquery) -> QueryPlan
+ * The tree traverse APIs like `transform`, `foreach`, `collect`, etc. that are
```

Comment:
@cloud-fan, does this mean all the APIs under `QueryPlan` and its downstream can go through query plans inside subqueries? It sounds like it.

Reply:
Okay, I just read the doc on each method you added. Should we mention what can go through and what can't?

Reply:
Only the `collectInPlanAndSubqueries` can go through subqueries, and it's documented.

Reply:
`transformExpressions` is not a traverse API as it only touches the current node.